src/algorithms/highlevel/crosssimilaritymatrix.cpp

```diff
@@ -0,0 +1,233 @@
+    heavisideStepFunction(similarityX);
+    heavisideStepFunction(similarityY);
+
+    csm.assign(similarityX.size(), std::vector<Real>(similarityY.size(), 0));
+    // tranpose similarityY vector for dot product
+    std::vector<std::vector<Real> > tSimilarityY = transpose(similarityY);
+    // clear memory
+    std::vector<std::vector<Real> >().swap(similarityY);
+
+    // Finally we construct out cross similarity matrix by doing dot product
+    for (size_t x=0; x<similarityX.size(); x++) {
+        for (size_t y=0; y<tSimilarityY.size(); y++) {
+            csm[x][y] = dotProduct(similarityX[x], tSimilarityY[y]);
+        }
+    }
+}
+
+
+// Construct a stacked chroma embedding from an input chroma audio feature vector [TODO] future use beat-synchronised stacked embeddings
+std::vector<std::vector<Real> > CrossSimilarityMatrix::toTimeEmbedding(std::vector<std::vector<Real> >& inputArray, int m, int tau) const {
+
+    int stopIdx;
+    int increment = m*tau;
+    int frameSize = inputArray[0].size() - increment;
+    int yDim = inputArray.size() * m;
+    std::vector<std::vector<Real> > timeEmbedding(frameSize, std::vector<Real>(yDim, 0));
+    std::vector<Real> tempRow;
+    tempRow.reserve(yDim*tau);
+
+    for (int i=0; i<frameSize; i+=tau) {
+        stopIdx = i + increment;
+        for (int startTime=i; startTime<=stopIdx; startTime+=tau) {
+            if (startTime == i) {
+                tempRow = inputArray[startTime];
+            }
+            else {
+                tempRow.insert(tempRow.end(), inputArray[startTime].begin(), inputArray[startTime].end());
+            }
+        timeEmbedding[i] = tempRow;
+        };
+    }
+    return timeEmbedding;
+}
+
+
+// computes global averaged chroma hpcp as described in [1]
+std::vector<Real> CrossSimilarityMatrix::globalAverageChroma(std::vector<std::vector<Real> >& inputFeature) const {
+
+    std::vector<Real> globalChroma;
+    std::vector<std::vector<Real> > featureCopy;
+    // tranpose the array from (time_axis, chroma_bin) to (chroma_bin, time-axis)for easy calculation
+    featureCopy = transpose(inputFeature);
+
+    // sum all the value along the time axis of chroma hpcp feature
+    for (size_t i=0; i<featureCopy.size(); i++) {
+        globalChroma.push_back(std::accumulate(featureCopy[i].begin(), featureCopy[i].end(), 0));
+    }
+
+    double maxItem = *std::max_element(globalChroma.begin(), globalChroma.end());
+    // divide the sum array by the max element to normalise it to 0-1 range
+    for (size_t j=0; j<globalChroma.size(); j++) {
+        globalChroma[j] = globalChroma[j] / maxItem;
+    }
+    return globalChroma;
+}
+
+
+// Compute the optimal transposition index for transposing reference song feature to the musical key of query song feature as described in [1].
+int CrossSimilarityMatrix::optimalTranspositionIndex(std::vector<std::vector<Real> >& featureA, std::vector<std::vector<Real> >& featureB, int nshifts) const {
```

Comment:
It makes sense to rename ``featureA`` and ``featureB`` to ``chromaA`` and ``chromaB`` for consistency with other variable names.